models/oauth2_application.go

```diff
@@ -0,0 +1,200 @@
+	if err != nil {
+		return "", err
+	}
+	app.ClientSecret = string(hashedSecret)
+	if _, err := x.ID(app.ID).Cols("client_secret").Update(app); err != nil {
+		return "", err
+	}
+	return secret, nil
+}
+
+// ValidateClientSecret validates the given secret by the hash saved in database
+func (app *OAuth2Application) ValidateClientSecret(secret []byte) bool {
+	return bcrypt.CompareHashAndPassword([]byte(app.ClientSecret), secret) == nil
+}
+
+// GetGrantByUserID returns a OAuth2Grant by its user and application ID
+func (app *OAuth2Application) GetGrantByUserID(userID int64) (*OAuth2Grant, error) {
+	return app.getGrantByUserID(x, userID)
+}
+
+func (app *OAuth2Application) getGrantByUserID(e Engine, userID int64) (grant *OAuth2Grant, err error) {
+	grant = new(OAuth2Grant)
+	if has, err := e.Where("user_id = ? AND application_id = ?", userID, app.ID).Get(grant); err != nil {
+		return nil, err
+	} else if !has {
+		return nil, nil
+	}
+	return grant, nil
+}
+
+// CreateGrant generates a grant for an user
+func (app *OAuth2Application) CreateGrant(userID int64) (*OAuth2Grant, error) {
+	return app.createGrant(x, userID)
+}
+
+func (app *OAuth2Application) createGrant(e Engine, userID int64) (*OAuth2Grant, error) {
+	grant := &OAuth2Grant{
+		ApplicationID: app.ID,
+		UserID:        userID,
+	}
+	_, err := e.Insert(grant)
+	if err != nil {
+		return nil, err
+	}
+	return grant, nil
+}
+
+// GetOAuth2ApplicationByClientID returns the oauth2 application with the given client_id. Returns an error if not found.
+func GetOAuth2ApplicationByClientID(clientID string) (app *OAuth2Application, err error) {
+	return getOAuth2ApplicationByClientID(x, clientID)
+}
+
+func getOAuth2ApplicationByClientID(e Engine, clientID string) (app *OAuth2Application, err error) {
+	app = new(OAuth2Application)
+	has, err := e.Where("client_id = ?", clientID).Get(app)
+	if !has {
+		return app, ErrOAuthClientIDInvalid{ClientID: clientID}
+	}
+	return
+}
+
+// CreateOAuth2Application inserts a new oauth2 application
+func CreateOAuth2Application(name string, userID int64) (*OAuth2Application, error) {
+	return createOAuth2Application(x, name, userID)
+}
+
+func createOAuth2Application(e Engine, name string, userID int64) (*OAuth2Application, error) {
+	clientID := gouuid.NewV4().String()
+	app := &OAuth2Application{
+		UID:          userID,
+		Name:         name,
+		ClientID:     clientID,
+		RedirectURIs: []string{},
+	}
+	if _, err := e.Insert(app); err != nil {
+		return nil, err
+	}
+	return app, nil
+}
+
+//////////////////////////////////////////////////////
+
+// OAuth2AuthorizationCode is a code to obtain an access token in combination with the client secret once. It has a limited lifetime.
+type OAuth2AuthorizationCode struct {
+	ID          int64        `xorm:"pk autoincr"`
+	Grant       *OAuth2Grant `xorm:"-"`
+	GrantID     int64
+	Code        string `xorm:"INDEX unique"`
+	RedirectURI string
+	ValidUntil  util.TimeStamp `xorm:"index"`
+}
```

Comment:
I am not sure this is a good implementation performance-wise. One trick that I learned from 
Microsoft implementations is that the access tokens don't have to be in the database at all, only the refresh tokens are stored there (and even that can be avoided to large extent). The idea is that the access token itself contains information about its scope and expiration, which is merely cryptographically signed and then base64 encoded. Validation of such token can be done purely through checking the cryptographic signature, which is only CPU bound and not storage bound. Naturally there's a problem that such token couldn't be easily revoked, but most implementations simply impose low limit for their lifetime (eg. 1 hour). Alternatively it's possible to implement revocation by simple in-memory structure.

Reply:
@lafriks I think he means JWT as alternative to access tokens we're already using.

Reply:
I've implemented it for refresh tokens too, since they are bound the the grant the user gave the application. If the user revokes this grant, the refresh token will become invalid too.